add autoscaling list command

chaozbj opened the pull request:
The issue is: #15 
here is an example of output:
```
12:00 ➜ ~/Workspace/code/github.com/kn-plugin-admin/cmd [chao-15 ✘] ./cmd autoscaling list
NAME                         VALUE
enable-scale-to-zero         false
max-scale-down-rate          3
panic-threshold-percentage   100
```

zhanggbj commented on pkg/command/autoscaling/list.go:
```diff
@@ -0,0 +1,101 @@
+	rows := make([]metav1beta1.TableRow, 0, len(configNameList))
+	for _, name := range configNameList {
+		row := metav1beta1.TableRow{}
+		if value := cm.Data[name]; value != "" {
```
@chaozbj one question, If there's no custom settings, cm.Data will be empty, and the user will get an empty output?
However, in this case, I think there're default autoscaling configs.

chaozbj replied:
Yes, nothing will be outputted, do you want to output default values? I'm not sure if the default value will be changed in the future in knative side, if changed, we need to update our plugin in time.

zhanggbj replied:
@chaozbj I think we should print the default values which could be found from Serving api/config map/or some where, not hardcoded in kn-admin.
As list command is to show the current configs to give users an overview, I think it's good to list default configs (only the global configs for admin)

CC @maximilien @rhuss What do you think? Thanks!

zhanggbj reviewed (commented):
@chaozbj it's good to add at least one e2e test case for `autoscaling list` in `test/e2e/kn_admin_test.go`, thanks!

maximilien commented on pkg/command/autoscaling/list.go:
```diff
@@ -0,0 +1,101 @@
+			currentCm := &corev1.ConfigMap{}
+			currentCm, err := p.ClientSet.CoreV1().ConfigMaps(knativeServing).Get(configAutoscaler, metav1.GetOptions{})
+			if err != nil {
+				return fmt.Errorf("failed to get ConfigMaps: %+v", err)
```
I wonder if putting the error message here will be useful to user?

maximilien commented on pkg/command/autoscaling/list_test.go:
```diff
@@ -0,0 +1,95 @@
+func TestAutoscalingListEmpty(t *testing.T) {
+	t.Run("no flags", func(t *testing.T) {
+		cm := &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
```
Maybe when it’s empty the result should be “No autoscaling info available” or similar instead of a list with headers...

chaozbj replied:
Thanks @maximilien, I will improve the output for empty configs.

chaozbj replied:
@maximilien and @zhanggbj Thanks, for e2e test case, yes, I missed it, I will add, but I saw @zhanggbj added the test function for autoscaling `func (et *e2eTest) knAdminAutoscaling(t *testing.T, r *test.KnRunResultCollector)` in her RP #16 , I think I should wait the PR to deliver first and then modify the same function to add tests for `list` command or I can add it later.

zhanggbj replied:
@chaozbj make sense, so let's wait #16 merge first and then you can rebase and add e2e in this PR then.

/hold

maximilien reviewed (approved):
Couple comments. Also need e2e tests. These are the best :)

Otherwise LGTM

zhanggbj commented:
/hold
wait for #16 to merge first.

zhanggbj commented:
/unhold
#16 is merged, please help to rebase and finish this PR, thanks!

chaozbj commented:
@zhanggbj @maximilien Changes as below:
1. Following `knative/client v0.17` to update `go.mod` because I want to use `autoscaler.Config.ScaleToZeroPodRetentionPeriod` field which is introduced from knative/serving v0.15. 

    I didn't follow the `knative/client v0.18 or v0.19` to update `go.mod` since I got compiling error when I use `k8s.io/apimachinery >= v0.18`, there is an interface changed from v0.18(maybe from earlier version): 
    ```go
    ../pkg/command/utils/utils.go:27:72: not enough arguments in call to client.CoreV1().ConfigMaps(desiredCm.ObjectMeta.Namespace).Get
	have (string, "k8s.io/apimachinery/pkg/apis/meta/v1".GetOptions)
	want (context.Context, string, "k8s.io/apimachinery/pkg/apis/meta/v1".GetOptions)
    ```
    We need to change our codes if we want to adopt it, but I don't want to address them in this PR, so I choose the older version to use.

2. Always list all values for autoscaler configs, if the config is changed by user, the new value will be printed, otherwise the default value will be printed. I called  `autoscalerconfig.NewConfigFromMap(map[string]string)` in knative/serving to build configs with default value, that is why I need to update `go.mod`
3. Updated unit test codes for new changes.
4. Added e2e test case for `autoscaling list` command.

Please help me to review, thanks!

zhanggbj commented on go.mod:
```diff
@@ -3,15 +3,22 @@ module knative.dev/kn-plugin-admin
+replace (
+	k8s.io/api => k8s.io/api v0.17.6
+	k8s.io/apimachinery => k8s.io/apimachinery v0.17.6
+	k8s.io/client-go => k8s.io/client-go v0.17.6
```
I suggest we finish dependency bump in #21 and then rebase on it

chaozbj replied:
ok, no problem.

zhanggbj commented:
@chaozbj Thanks Chao for the efforts! In general, the changes look good to me, 

For 1) Let's finish the dependency bump firstly in PR https://github.com/knative-sandbox/kn-plugin-admin/pull/21, we need it to build release 0.1.0,  this will also resolve the issue you mentioned.

For 5) I think `./hack/xx` is more proper for some scripts and tools like release.sh

chaozbj commented on cmd/release.sh:
```diff
@@ -25,11 +25,11 @@ export CGO_ENABLED=0
 
 # build for macOS
```
@zhanggbj Thanks, I will take a look and move `release.sh` to here

chaozbj commented:
@zhanggbj I rebased from PR #25 and also delete `release.sh` script because I found that `hack/build.sh` can do the same thing. Please help me review, thanks!

zhanggbj commented:
/lgtm
Look nice and merging! Thanks @chaozbj!

zhanggbj commented:
Prow tests are failing due to test-infra setting in progress which is covered by https://github.com/knative-sandbox/kn-plugin-admin/pull/24
```
/usr/local/bin/kubekins-runner.sh: line 104: ./test/presubmit-tests.sh: No such file or directory
```

maximilien reviewed (commented):
Tests still failing
/retest

navidshaikh commented:
/test all

zhanggbj commented:
/retest

zhanggbj commented:
/lgtm cancel

@chaozbj would you please help to rebase first. Just merged Navid's PR for test infra.

chaozbj commented:
/retest

chaozbj commented:
/retest

navidshaikh commented:
@chaozbj : there is a verify codegen step now part of the build tests,
```

ERROR: /home/prow/go/src/knative.dev/kn-plugin-admin is out of date. Please run ./hack/build.sh -c and commit.
--

please run `./hack/build.sh` and commit the changes

chaozbj commented:
close it and will create a new PR for `autoscaling list` command
